app/tasks/email_notifications.rb

```diff
@@ -66,14 +64,21 @@ def feedback_ending(time_text)
                        .pluck(:user_id))
 
     Ajat.info "feedback_ending|id:#{@contest.id}|time:#{time_text}"
-    send_emails(text: text, subject: subject, users: users)
+    send_emails(text: text, subject: subject, users: users, event: 'feedback_ending')
   end
 
   private def send_emails(**hash)
-    hash[:users].pluck(:email).each_slice(200) do |arr|
+    hash[:users].pluck(:email).each do |email|
+      user_id = User.find_by(email: email).id
+      notification_id = Notification.find_by(event: hash[:event]).id
+      user_notification = UserNotification.where(user_id: user_id).find_by(notification_id: notification_id)
+      unsubscribe_link = unsubscribe_url token: user_notification.generate_token
```

Comment:
I think a better design would be to generate the token automatically before UserNotification create. That way we can set UserNotification to be non-nullable, and we can just do `user_notification.token` here instead of `user_notification.generate_token`.

That way, when someone else wants to use the UserNotification's token, they would not need to do the hassle of generating the tokens themselves, as it would have been auto generated.